pkg/resource/training_job/custom_set_output.go

```diff
@@ -0,0 +1,107 @@
+// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//     http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+// Use this file if the Status/Spec of the CR needs to be modified after
+// create/describe/update operation
+
+package training_job
+
+import (
+	"context"
+
+	ackv1alpha1 "github.com/aws-controllers-k8s/runtime/apis/core/v1alpha1"
+	svcapitypes "github.com/aws-controllers-k8s/sagemaker-controller/apis/v1alpha1"
+	"github.com/aws/aws-sdk-go/aws"
+	svcsdk "github.com/aws/aws-sdk-go/service/sagemaker"
+	corev1 "k8s.io/api/core/v1"
+)
+
+// customCreateTrainingJobSetOutput sets the resource in TempOutofSync if TrainingJob is
+// in creating state. At this stage we know call to createTrainingJob was successful.
+func (rm *resourceManager) customCreateTrainingJobSetOutput(
+	ctx context.Context,
+	r *resource,
+	resp *svcsdk.CreateTrainingJobOutput,
+	ko *svcapitypes.TrainingJob,
+) (*svcapitypes.TrainingJob, error) {
+	rm.customSetOutput(r, aws.String(svcsdk.TrainingJobStatusInProgress), ko)
+	return ko, nil
+}
+
+// customDescribeTrainingJobSetOutput sets the resource in TempOutofSync if
+// TrainingJob is being modified by AWS. It has an additional check on the debugger status.
+func (rm *resourceManager) customDescribeTrainingJobSetOutput(
+	ctx context.Context,
+	r *resource,
+	resp *svcsdk.DescribeTrainingJobOutput,
+	ko *svcapitypes.TrainingJob,
+) (*svcapitypes.TrainingJob, error) {
+	trainingJobStatus := resp.TrainingJobStatus
+	debuggerRuleInProgress := false
+	if resp.DebugRuleEvaluationStatuses != nil {
+		for _, rule := range resp.DebugRuleEvaluationStatuses {
+			if rule.RuleEvaluationStatus != nil && *rule.RuleEvaluationStatus == svcsdk.RuleEvaluationStatusInProgress {
+				debuggerRuleInProgress = true
+				rm.customSetOutput(r, aws.String(svcsdk.TrainingJobStatusInProgress), ko)
+				break
+			}
+		}
+	}
+
+	if !debuggerRuleInProgress {
+		rm.customSetOutput(r, trainingJobStatus, ko)
+	}
```

Comment:
Why are we doing this client-side evaluation of the training job status based on a "debug rule evaluation"? Shouldn't the SageMaker service-side do this determination?